### DISCUSSION TITLE
How to debug dbeaver with idea

### DISCUSSION THREAD

**Discussion by taohaozhi1129** [dbeaver/dbeaver#36053] [DBeaver Development]
How to debug dbeaver with idea, I saw the official website documentation is not rich, can you have more detailed documentation
Thanks
![image](https://github.com/user-attachments/assets/58326f87-5f5d-4d7b-bb3b-4beb803f9648)

**Comment by flambchap** [comment]
Hi, @taohaozhi1129. I see that a couple of important repos are missing in your workspace. Please follow this guide to configure IntelliJ IDEA's workspace: https://github.com/dbeaver/dbeaver/wiki/Develop-in-IDEA.

**Comment by taohaozhi1129** [comment]
@n5a5 Hello, I follow the document, I finished step 6 directory structure is like this, The following steps do not understand the specific how to do, can you help me specifically how to do? What commands or scripts should I execute, please
![image](https://github.com/user-attachments/assets/d1e44757-b129-4309-9688-5b1d893149a5)
![image](https://github.com/user-attachments/assets/85711052-57e2-4ec7-9201-898a0a75fe09)


> Hi, @taohaozhi1129. I see that a couple of important repos are missing in your workspace. Please follow this guide to configure IntelliJ IDEA's workspace: https://github.com/dbeaver/dbeaver/wiki/Develop-in-IDEA.

**Comment by flambchap** [comment]
7. Execute `dbeaver/generate_workspace.cmd`
8. The script above will create the `dbeaver-workspace/dbeaver-ce` folder. Open it as an IntelliJ IDEA project.
9. Follow step 9 as in https://github.com/dbeaver/dbeaver/wiki/Develop-in-IDEA#opening-dbeaver-in-idea
10. `mvn clean compile -T 1C -f dbeaver/product/aggregate`
11. Compile the workspace

**Comment by taohaozhi1129** [comment]
@n5a5 At present, dbeaver can be started and debugged normally in idea, thank you very much for your help。
I have another question. For instance, if I want to develop a database connection plugin, what development standards should I follow? Are there any references for this? Additionally, how do I package it after development is completed?
I noticed that drivers are usually downloaded online. If I have some special driver packages, can I package them directly so that they are available for use when creating a new connection, without needing to download them again?
![image](https://github.com/user-attachments/assets/bfd116e2-b13c-47db-9b6c-593babfdaab8)

**Reply by Learng1405** [reply]
Hi, after completing all steps till 10, how do i run the program in intelliJ, as it is asking main class to run the application, i am unable to find main class.

**Reply by flambchap** [reply]
Hi. `mvn clean install` for [this](https://github.com/dbeaver/dbeaver/tree/devel/plugins/org.jkiss.dbeaver.launcher) plugin should help

**Reply by Learng1405** [reply]
![Screenshot from 2024-11-14 11-03-05-1](https://github.com/user-attachments/assets/cdb406f9-e511-443c-a246-c3c5e13ef935)
@n5a5 =,  
done that step, but how to start application using IntelliJIDea or by some binary file. 
give steps
when i am trying to do by IntelliJIdea it is asking about the main class to run the application, what is the main class

**Reply by flambchap** [reply]
I'm sorry for the late response. I totally missed your reply somehow :(

To fix the issue, please open the `dbeaver-ce` folder as an IDEA project, not the `dbeaver-workspace` folder.

**Comment by Destrolaric** [comment]
Hi we will convert it to discussion for now, there are a lot of things what might differ between each driver implementation. Documentation unfortunately will not help with that.

**Comment by flambchap** [comment]
@taohaozhi1129 This document might interest you: https://github.com/dbeaver/dbeaver/wiki/Contribute-your-code. When it comes to adding support for other databases, I suggest using previous pull requests as a reference. You can find them by looking at associated tickets on our issue tracker with the `new driver` label: https://github.com/dbeaver/dbeaver/issues?q=is%3Aissue+label%3A%22new+driver%22+is%3Aclosed. 

Unfortunately, we won't accept a pull request with an implementation that directly packages the driver library. If we packaged DBeaver with each driver, its installation size would grow enormously.

**Reply by taohaozhi1129** [reply]
@n5a5  Okay, thank you. I will first take a look at the relevant links you provide and consult you if I have any questions

**Reply by taohaozhi1129** [reply]
@n5a5 Hello, I successfully compiled and built it from source, but I encountered an issue. The startup requires Java environment support, but during the build process, the Java environment is not included.
![image](https://github.com/user-attachments/assets/45f2354c-84d3-4bcd-b140-09424ac00fea)

**Reply by flambchap** [reply]
This is expected. I can think of 3 solutions:

1. Make Java 17 the default Java on your system
2. Download and unpack JRE 17 right into this directory
3. My favorite: provide the path to Java 17 in dbeaver.ini using the -vm flag before the -vmargs:

```
...
-vm
/path/to/jvm/17/on/your/machine
-vmargs
...
```

**Reply by taohaozhi1129** [reply]
Thank you, the method above was useful

**Reply by taohaozhi1129** [reply]
@n5a5 I noticed that each time a new database connection is created, the required driver can only be downloaded from the internet. For example, when developing new connectors, I have some self-developed driver packages that aren’t available online yet. Is it possible to embed these drivers during packaging?
![image](https://github.com/user-attachments/assets/04b90c02-94b2-48c7-a4b8-f7fb115870d4)

**Reply by flambchap** [reply]
It is possible, but as I mentioned previously, we won't merge a feature request with a code that embeds a database driver.
